azure-iot-device/tests/common/pipeline/test_pipeline_stages_base.py

```diff
@@ -715,3 +718,460 @@ def test_ignores_unknown_request_id(self, stage, op, iot_response, unhandled_err
+        assert getattr(yes_retry_op, "retry_timer", None) is None
+
+
+class RetryStageTestResubmitedOpCompletion(object):
+    """
+    Tests for RetryStage for resubmitted op completion
+    """
+
+    @pytest.mark.it("Calls the original callback with success when the retried op succeeds")
+    def test_calls_callback_on_retried_op_success(
+        self, stage, yes_retry_op, retry_error, mock_timer
+    ):
+        op_callback = yes_retry_op.callback
+        stage.run_op(yes_retry_op)
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        timer_callback = mock_timer.call_args[0][1]
+        timer_callback()
+        assert op_callback.call_count == 0
+        stage.next._complete_op(op=yes_retry_op)
+        assert yes_retry_op.callback == op_callback
+        assert_callback_succeeded(op=yes_retry_op)
+
+    @pytest.mark.it(
+        "Calls the original callback with error when the retried op compltes with an arbitrary error"
+    )
+    def test_calls_callback_on_retried_op_arbitrary_exception(
+        self, stage, yes_retry_op, retry_error, mock_timer, arbitrary_exception, mocker
+    ):
+
+        stage.run_op(yes_retry_op)
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        timer_callback = mock_timer.call_args[0][1]
+        timer_callback()
+        stage.next._complete_op(op=yes_retry_op, error=arbitrary_exception)
+        assert_callback_failed(op=yes_retry_op, error=arbitrary_exception)
+
+    @pytest.mark.it(
+        "Does not calls the original callback with error when the retried op compltes with an retry error"
+    )
+    def test_no_callback_on_retried_op_retry_error(
+        self, stage, yes_retry_op, retry_error, mock_timer
+    ):
+        op_callback = yes_retry_op.callback
+        stage.run_op(yes_retry_op)
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        timer_callback = mock_timer.call_args[0][1]
+        timer_callback()
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        assert op_callback.call_count == 0
+
+    @pytest.mark.it("Sets a new retry timer error when the retried op compltes with an retry error")
+    def test_sets_timer_on_retried_op_retry_error(
+        self, stage, yes_retry_op, retry_error, mock_timer
+    ):
+        stage.run_op(yes_retry_op)
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        assert mock_timer.call_count == 1
+        timer_callback = mock_timer.call_args[0][1]
+        timer_callback()
+        stage.next._complete_op(op=yes_retry_op, error=retry_error)
+        assert mock_timer.call_count == 2
+
+
+@pytest.mark.describe("RetryStage - run_op()")
+class TestRetryStageRunOp(
+    StageTestBase,
+    RetryStageTestOpSend,
+    RetryStageTestNoRetryOpCallback,
+    RetryStageTestNoRetryOpSetTimer,
+    RetryStageTestYesRetryOpCallback,
+    RetryStageTestYesRetryOpSetTimer,
+    RetryStageTestResubmitOp,
+    RetryStageTestResubmitedOpCompletion,
+):
```

Comment:
This is an interesting model. Not sure I have anything constructive to add, but it's something I haven't thought of, and I'll take this into consideration in the future (including the stuff I'm working on now). Any particular motivations or pros and cons you'd like to highlight?

Reply:
I did it for logical organization.  I could have just as easily have done it by grouping tests in a single class.

Pros:
* Forced me to think about logical organization of my tests.
* Forced me to think about missed cases ("wait, I have 3 tests for this kind of op, I think I need the same 3 tests for that other kind of op")
* Makes the logical organization to be more explicit and more easily understood (I hope)
* Encourages future developers (including future me) to think about these same points
Cons:
* Relies on other classes (parents and siblings) to define fixtures and helpers (see the `retry_error` fixture above)
* Bypasses some static analysis checks.   e.g. no warning if RetryStageTestYesRetryOpSetTimer and RetryStageTestYesRetryOpSetTimer both define tests with the same name.  Only one will run.


---
In reply to: [337214623](https://github.com/Azure/azure-iot-sdk-python/pull/365#discussion_r337214623) [](ancestors = 337214623)